File: android-navigator/src/main/java/com/github/kostasdrakonakis/androidnavigator/IntentNavigatorBinder.java
package com.github.kostasdrakonakis.androidnavigator;

import android.app.Activity;
import android.util.Log;

import androidx.annotation.CheckResult;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.UiThread;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.LinkedHashMap;
import java.util.Map;

public final class IntentNavigatorBinder {

    private static final String TAG = IntentNavigatorBinder.class.getSimpleName();
    private static boolean debug;

    private static final Map<Class<?>, Constructor> CONSTRUCTOR_MAP = new LinkedHashMap<>();

    private IntentNavigatorBinder() {
        throw new UnsupportedOperationException("No instances allowed");
    }

    @UiThread
    public static void bind(@NonNull Activity activity) {
        createBinder(activity);
    }

    public static void setDebug(boolean debug) {
        IntentNavigatorBinder.debug = debug;
    }

    private static void createBinder(@NonNull Activity activity) {
        Class<?> targetClass = activity.getClass();
        try {
            Constructor constructor = getConstructorForClass(targetClass);
            if (constructor != null) {
                constructor.newInstance(activity);
            }
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (InstantiationException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        }
    }

    @Nullable
    @CheckResult
    @UiThread
    private static Constructor getConstructorForClass(Class<?> cls) {
        Constructor<?> bindingConstructor = CONSTRUCTOR_MAP.get(cls);
        if (bindingConstructor != null) {
            if (debug) Log.d(TAG, "Returned constructor from Cache.");
            return bindingConstructor;
        }

        String clsName = cls.getPackage().getName() + "." + cls.getSimpleName();
        if (clsName.startsWith("android.") || clsName.startsWith("java.")) {
            if (debug) Log.d(TAG, "Reached framework class. No further.");
            return null;
        }

        try {
            Class<?> bindingClass =
                    cls.getClassLoader().loadClass(clsName + "_INTENT_PROPERTY_BINDING");
            if (debug) Log.d(TAG, "Binding Class: " + bindingClass);

            bindingConstructor = bindingClass.getConstructor(cls);
            if (debug) Log.d(TAG, "Loaded binding class and constructor.");
        } catch (ClassNotFoundException e) {
            return null;
        } catch (NoSuchMethodException e) {
            throw new RuntimeException("Unable to find binding constructor for " + clsName, e);
        }
        CONSTRUCTOR_MAP.put(cls, bindingConstructor);
        return bindingConstructor;
    }
}

File: android-navigator/src/main/java/com/github/kostasdrakonakis/androidnavigator/IntentNavigatorBinder.kt
package com.github.kostasdrakonakis.androidnavigator

import android.app.Activity
import android.util.Log
import androidx.annotation.CheckResult
import androidx.annotation.UiThread
import java.lang.reflect.Constructor
import java.lang.reflect.InvocationTargetException

object IntentNavigatorBinder {
    private val TAG = IntentNavigatorBinder::class.java.simpleName
    private var debug = false
    private val CONSTRUCTOR_MAP: MutableMap<Class<*>, Constructor<*>> = linkedMapOf()

    @JvmStatic
    @UiThread
    fun bind(activity: Activity) {
        createBinder(activity)
    }

    fun setDebug(debug: Boolean) {
        this.debug = debug
    }

    private fun createBinder(activity: Activity) {
        val targetClass: Class<*> = activity.javaClass
        try {
            val constructor = getConstructorForClass(targetClass)
            constructor?.newInstance(activity)
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        } catch (e: InstantiationException) {
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        }
    }

    @CheckResult
    @UiThread
    private fun getConstructorForClass(cls: Class<*>): Constructor<*>? {
        var bindingConstructor = CONSTRUCTOR_MAP[cls]
        if (bindingConstructor != null) {
            if (debug) Log.d(TAG, "Returned constructor from Cache.")
            return bindingConstructor
        }
        val clsName = cls.getPackage()!!.name + "." + cls.simpleName
        if (clsName.startsWith("android.") || clsName.startsWith("java.")) {
            if (debug) Log.d(TAG, "Reached framework class. No further.")
            return null
        }
        try {
            val bindingClass = cls.classLoader!!.loadClass(clsName + "_INTENT_PROPERTY_BINDING")
            if (debug) Log.d(TAG, "Binding Class: $bindingClass")
            bindingConstructor = bindingClass.getConstructor(cls)
            if (debug) Log.d(TAG, "Loaded binding class and constructor.")
        } catch (e: ClassNotFoundException) {
            return null
        } catch (e: NoSuchMethodException) {
            throw RuntimeException("Unable to find binding constructor for $clsName", e)
        }
        CONSTRUCTOR_MAP[cls] = bindingConstructor
        return bindingConstructor
    }
}